core/src/main/scala/org/apache/spark/rdd/OrderedRDDFunctions.scala

```diff
@@ -73,7 +75,21 @@ class OrderedRDDFunctions[K : Ordering : ClassTag,
    * because it can push the sorting down into the shuffle machinery.
    */
   def repartitionAndSortWithinPartitions(partitioner: Partitioner): RDD[(K, V)] = self.withScope {
-    new ShuffledRDD[K, V, V](self, partitioner).setKeyOrdering(ordering)
+    if (self.partitioner == Some(partitioner)) {
+      self.mapPartitions(iter => {
+        val context = TaskContext.get
+        val sorter = new ExternalSorter[K, V, V](context, None, None, Some(ordering))
+        sorter.insertAll(iter)
+        context.taskMetrics.incDiskBytesSpilled(sorter.diskBytesSpilled)
+        context.taskMetrics.incMemoryBytesSpilled(sorter.memoryBytesSpilled)
+        context.taskMetrics.incPeakExecutionMemory(sorter.peakMemoryUsedBytes)
+        val outputIter = new InterruptibleIterator(context,
+          sorter.iterator.asInstanceOf[Iterator[(K, V)]])
+        CompletionIterator[(K, V), Iterator[(K, V)]](outputIter, sorter.stop)
+      }, preservesPartitioning = true)
```

Comment:
Can we refactor the code here and `BlockStoreShuffleReader.read` ?

Reply:
@mridulm sorry for the late reply.
I think I am missing something, I don't quite understand why/how to refactor here, do you mean adding a Listener like this?
```
        // Use completion callback to stop sorter if task was finished/cancelled.
        context.addTaskCompletionListener[Unit](_ => {
          sorter.stop()
        })
```

Reply:
current pr is also similar to `def distinct(numPartitions: Int)` in `RDD.scala`, I think the main difference is that : in `distinct(numPartitions: Int)` a ExternalAppendOnlyMap is used.